File: 0078-subsets/0078-subsets.py
class Solution:
    def printSubsets(self,ans,arr,ind,temp):
        if ind==len(arr):
            ans.append(temp[:])
            return
        self.printSubsets(ans,arr,ind+1,temp)
        temp.append(arr[ind])
        self.printSubsets(ans,arr,ind+1,temp)
        temp.pop()

    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans=[]
        self.printSubsets(ans,nums,0,[])
        return ans

File: 0078-subsets/0078-subsets.java
class Solution {
    public void gen(int ind,List<List<Integer>> ans,int[] arr,List<Integer> dup){
        if(ind==arr.length){
            ans.add(new ArrayList<>(dup));
            return;
        }
        dup.add(arr[ind]);
        gen(ind+1,ans,arr,dup);
        dup.remove(dup.size()-1);
        gen(ind+1,ans,arr,dup);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans=new ArrayList<>();
        List<Integer> dup= new ArrayList<>();
        gen(0,ans,nums,dup);
        return ans;
    }
}